[Shipping] Unformat Shipping Method numeric values before sending data

psealock opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/53876
Closes https://github.com/woocommerce/woocommerce/issues/53315
Closes https://github.com/woocommerce/woocommerce/issues/47795

When Shipping Method values were edited in the the Shipping Method Modal, numbers with thousandth and decimal separators were sent as human readable strings with separators. This caused errors on the server and values were getting saved as `0.00`. This PR makes sure the formatting separators are removed and the value sent to the server is a valid number.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to WooCommerce->Settings->Shipping->(any zone that you are in)
2. Set up shipping method: with type flat rate
3. Enter 1000 in the cost field and click "save"
4. Once again, edit the shipping method by clicking "Edit" and saving. This step is important as this is where the error occurs on trunk.
5. Add a test product with price, enable a payment method like Cash on delivery
6. On the shopper side, add the product to your cart and note the shipping method and its price are accurately shown.

<img width="553" alt="image" src="https://github.com/user-attachments/assets/8e5e5b34-7813-406c-8597-1a113d2a9718" />

7. Test a formula as the value by placing `1000 * [qty]` as the value and save. In the cart, refresh the page and increase the product quantity in your cart. The flat rate shipping should update accordingly.
8. Repeat this with different thousand and decimal separators at WooCommerce > Settings > General > Thousand separator | Decimal separator.
9. Try once again with free shipping and a minimum order amount.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Unformat Shipping Method numeric values before persisting data to the server.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

psealock committed:
Unformat shipping method numeric values before sending to server

psealock committed:
add comment

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

psealock committed:
handle formulas

rjchow reviewed (dismissed):
This doesn't fix the case where the merchant enters the comma and decimal point intentionally - e.g a min order amount of "1,000.00" will not be correctly applied.

I think let's fix this properly by both ensuring that the frontend only sends a valid numeric value without thousands separator, currency symbol, and a period for indicating decimal points, as well as ensuring that the backend properly fails when an invalid value according to those rules is sent. Configuration that includes formulae should be separately handled, ideally in another endpoint. This should do its own validation, since there's really no overlap. 

A more risky follow up would be to also correctly fail when an invalid value is encountered when evaluating shipping costs. This is risky because it will break existing non-compliant store setups. A way to make this less risky is to only flag it in a notice in WC Admin if it is detected on init.

Edit: for expediency we can just mitigate it in frontend first, which just needs an extension of this PR to also handle intentional entries. (I haven't looked into why it doesn't, but just entering a minimum free shipping criteria of "1,000.00" in the US currency settings gives away free shipping on a cart value of $10)

Edit 2: A small UI tweak to make validation easier would be to split the field into integer and float fields with the separator as a fixed part of the UI

rjchow commented:
> This doesn't fix the case where the merchant enters the comma and decimal point intentionally - e.g a min order amount of "1,000.00" will not be correctly applied.

More testing indicates that this was an inaccurate finding, it seems that the unformatting is not being triggered on the 'Rest of world' zone rather than the unformatting not applying at all for min order amount of "1,000.00"

waclawjacek commented on plugins/woocommerce/client/legacy/js/admin/wc-shipping-zone-methods.js:
```diff
@@ -351,6 +351,44 @@
+							return;
+						}
+
+						// Brackets signal a formula. Avoid unformatting these values.
```
Could we maybe have some logic to sanitize the string, split it into components (number/not number), and then run the sanitization on all numbers instead of not doing anything at all when there's a shortcode?

psealock replied:
>then run the sanitization on all numbers

Are you suggesting that we accept something like `[qty] * 4,5` for locales with commas as decimal places and it becomes `[qty] * 4.5` in the database? If so, I'd think this would be something outside the scope of this PR.

rjchow replied:
I agree it's outside the scope of this PR - and possibly entirely outside the scope of what we should support. There was a previous PR that tried to allow this but it got reverted due to the mess it made.

I do think we should define a grammar for the formula and then parse it according to the grammar though, that shouldn't be too much work and will be reuseable going forward.

rjchow replied:
(with just standard ISO decimal numbers in the formula)

waclawjacek reviewed (changes requested):
Thanks for this PR! 🙌 

I did some testing and found behavior that makes me think that we should have some more/different logic to handle the formatting.

I see that while WooCommerce doesn't allow using completely made-up values for the shipping cost (tested: `  foobar `), there's a lot you can get away with.

For example, I got the following results when testing with a shipping cost input of `1 000,00.00`:

<table>
<tr>
<th>Thousands separator</th>
<th>Decimal separator</th>
<th>What was sent in the HTTP update request (on this branch)</th>
</tr>
<tr><td><kbd>,</kbd></td><td><kbd>.</kbd></td><td>1 00000.00</td></tr>
<tr><td><kbd>,</kbd></td><td><kbd>&nbsp;</kbd></td><td>1.00000.00</td></tr>
<tr><td><kbd>&nbsp;</kbd></td><td><kbd>,</kbd></td><td>1000.00.00</td></tr>
<tr><td><kbd>&nbsp;</kbd></td><td><kbd>.</kbd></td><td>1000,00.00</td></tr>
</table>

I am also not super confident about having one behavior happen when a shortcode is not present (sanitize), and another - when it is present (no sanitization). I see the value is covering many cases that don't use a shortcode but it'd be great to have a fix for all cases.

Additionally, I also came across the same issue that @rjchow reported - the sanitization doesn't trigger for the "Rest of the world" shipping zone. It seems to be caused by `window.wc.ShippingCurrencyContext` not being defined in that zone.

I think an ideal solution for me (which I realize might not be viable to implement but just sharing in case it might be) would be if WooCommerce ran validation upon every attempt to store a number that has characters outside of the expected separators (looking only at the current settings, not all possible separators for all locales), and then persisted the value in the database as an integer instead of a value with separators (e.g. if the input is `5.50`, we could store `550`). This would also allow changing the separator settings without affecting the values (although this is probably a super edge case).

rjchow commented:
Hey @waclawjacek ! Thanks for the review - there's one bit to clarify 
> It seems to be caused by window.wc.ShippingCurrencyContext not being defined in that zone.

I think rather than that, it's because it has [its own saving routine](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/js/admin/wc-shipping-zone-methods.js#L36) which originally ran the same code, but was not kept in sync when more logic was added to `onConfigureShippingMethod`.

For the purpose of this PR I think I would be satisfied if we just extract the logic and call it from both of these functions so that this divergence doesn't happen again. Additionally some unit tests around the new `unformatShippingMethodNumericValues()` method would be prudent

rjchow commented on plugins/woocommerce/client/legacy/js/admin/wc-shipping-zone-methods.js:
```diff
@@ -351,6 +351,44 @@
+							return;
+						}
+
+						if ( 'number' === typeof formattedValue ) {
```
I think we should use `Number.isFinite()` to test instead - `NaN` also evalutes to typeof 'number'

waclawjacek commented:
I have prepared an alternative solution that only accepts numbers using only the thousand and decimal separators defined in settings, then saves them as numbers passing the `is_numeric()` check in the database.

Then for display, it converts them back to a format using separators defined in settings.

The PR is here: https://github.com/woocommerce/woocommerce/pull/54542

psealock committed:
Use window.wc.wcSettings.CURRENCY

psealock committed:
use Number.isFinite()

psealock commented:
>I think let's fix this properly by both ensuring that the frontend only sends a valid numeric value without thousands separator, currency symbol, and a period for indicating decimal points, as well as ensuring that the backend properly fails when an invalid value according to those rules is sent. Configuration that includes formulae should be separately handled, ideally in another endpoint. This should do its own validation, since there's really no overlap.

This seems like the proper solution. While making sure the client isn't sending improper strings as numbers solves the immediate concern, making sure the backend rejects invalid values future proofs the situation. @waclawjacek, what do you think? Seems to me like we should merge both approaches.

psealock commented:
>I think rather than that, it's because it has [its own saving routine](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/js/admin/wc-shipping-zone-methods.js#L36) which originally ran the same code, but was not kept in sync when more logic was added to onConfigureShippingMethod.

I was able to get the logic running for "Rest of the world" by using `window.wc.wcSettings.CURRENCY` instead of the context, which wasn't available. I can't get this `save` function to fire, so not sure where its actually used 🤷🏽 .

psealock committed:
remove all non numeric and separator characters

psealock commented:
>For example, I got the following results when testing with a shipping cost input of 1 000,00.00:

In [5fcf3b3](https://github.com/woocommerce/woocommerce/pull/54181/commits/5fcf3b3b5a50202d5652322aae9765df895352b6) I add a regex to strip the input of all non numeric and separator values. This helps with the `1 000,00.00` input scenario.

>Additionally some unit tests around the new unformatShippingMethodNumericValues() method would be prudent

I agree. We'll need to add some testing infrastructure

waclawjacek commented:
> Seems to me like we should merge both approaches.

One issue I can see with forcing front-end strings to be formatted like in the US locale is that it could be counterintuitive to merchants to input using commas and a dot if they are used to, say, spaces and a comma (like we are in Poland). What do you think would be the best solution here? In #54542 I made it so that only amounts formatted using the separators specified in WC settings are accepted, and then the backend will also look at what the separators are and convert the number into a number that passes PHP's `is_numeric` check.

waclawjacek commented:
> @waclawjacek the frontend display is still formatted according to the locale, the unformatting only happens pre-http request.

Doesn't it only happen the first time, after which it's US locale-formatted?

waclawjacek commented:
> I think it would be great to merge your approach into this PR if that's acceptable? I like that you've extracted the duplicated sanitisation code!

For sure! I'll do that now.

Commit:
Sanitize localized shipping method cost (#54542)

Only allow shipping method cost expressed using numbers and the thousand and decimal separators set in WC settings.

If a value is outside of that character set, display a notice.

Reformat these numbers to is_numeric() strings for DB persistence and then reformat them to localized strings for display.

rjchow committed:
added tests and changed some names for clarity

rjchow committed:
lint

rjchow commented:
@psealock , @waclawjacek I worked on the unit tests for the backend to move things along and found / fixed some issues 
- there was a sanitisation miss with wp_kses_post normalising the html entities, which made them not match the str_replace for the currency entities
- I realised that the root zone still had issues because of the reliance on the injected context which was not being injected, so i moved the formatting functions to woo/currency

After all that I realised that the new unformatting code quietly reformats things like $1ooo to $1 without erroring, so if the user makes a typo they won't realise. I think that's not great so I'm working on that next.

adrianduffell commented:
@rjchow Would we be able add the tests in a standalone PR? I envision that we would be able to see most tests passing with some needing to be skipped due to the present bugs. Then, with this PR, all tests would pass without any changes to the tests. I think it would help us prove this PR doesn't introduce regressions.

psealock commented on plugins/woocommerce/client/legacy/js/admin/wc-shipping-zone-methods.js:
```diff
@@ -457,12 +499,12 @@
 				},
 				addCurrencySymbol: function( html ) {
-					if ( ! window.wc.ShippingCurrencyContext || ! window.wc.ShippingCurrencyNumberFormat ) {
+					if ( ! window.wc.wcSettings.CURRENCY || ! window.wc.currency.localiseMonetaryValue ) {
```
@rjchow, optional chaining here?
```suggestion
					if ( ! window.wc.wcSettings.CURRENCY || ! window.wc.currency?.localiseMonetaryValue ) {
```

rjchow replied:
I initially started off with it but then I realised we don't transpile this file so it doesn't get polyfilled for lower envs. But I just looked up the browser support and I think it's fine so I'll make the change

<img width="1375" alt="image" src="https://github.com/user-attachments/assets/e958f511-675b-4510-9965-f2c6797383dc" />

psealock reviewed (commented):
>After all that I realised that the new unformatting code quietly reformats things like $1ooo to $1 without erroring, so if the user makes a typo they won't realise. I think that's not great so I'm working on that next.

Theres a `validateFormArguments` function on the class we can hook into. It runs validation on every input, which isn't as nice as on blur, but we can run ` Number.isFinite( formattedData[ field ] )` before enabling the save button. Maybe thats best in a follow up

psealock commented:
Noting here that we need to extend this logic to these class fields
<img width="523" alt="Screenshot 2025-01-21 at 11 08 46 PM" src="https://github.com/user-attachments/assets/93a516cc-fedd-46c8-a4b1-d113e5e94b43" />

rjchow committed:
improved ux by letting unrecoverable errors pass through

Commit:
Add changefile(s) from automation for the following project(s): @woocommerce/currency, woocommerce, woocommerce/client/admin

psealock commented on plugins/woocommerce/src/Utilities/NumberUtil.php:
```diff
@@ -49,4 +49,63 @@ public static function array_sum( array $arr ): float {
+
+		if ( 1 !== preg_match( $allowed_characters_regex, $value ) ) {
+			throw new \InvalidArgumentException( sprintf( __( '%s is not a valid numeric value. Allowed characters: %s', 'woocommerce' ), $value, $allowed_characters_regex ) );
+		}
```
This is great feedback when things go wrong! We should offer a more merchant friendly message though. How about something like: `1oooo is not a valid numeric value. Please make sure to enter a valid number`

<img width="648" alt="image" src="https://github.com/user-attachments/assets/4c33a79f-1682-40bf-bf59-07a734123e4b" />

rjchow replied:
I had something like that initially but it just seemed to provide less information and the second sentence doesn't help the user troubleshoot 🙈 I could translate the regex to english?

waclawjacek replied:
Not sure if that's helpful but we could also include the separators we're talking about in the message, like:

> [...] are the thousand (`,`) and decimal (`.`) separators.

psealock committed:
include shipping class inputs

psealock commented:
>Noting here that we need to extend this logic to these class fields

@rjchow I handled this in https://github.com/woocommerce/woocommerce/pull/54181/commits/28c66bb189659eebd43a4da5ed9c29730b911e40

#### To Test

1. Make sure you have a shipping class. Add two to fully test.
2. Go to or add a shipping zone with a flat rate shipping method
3. Play around with the shipping class costs inputs and see the formatting occur on blur. The formatting will be based on your settings.
4. In the Network tab, filter by `woocommerce_shipping_zone_methods_save_settings` and check the values sent by the client regarding shipping class costs. They should be stripped of their formatting in favour of a numeric value.

rjchow committed:
lint

rjchow committed:
tests

rjchow committed:
disable strict types lint for NumberUtil

rjchow committed:
Lint.

rjchow reviewed (approved):
Ok I think it's good to go, thanks! @waclawjacek would you like to give it a final look over?

waclawjacek commented:
I'll take a look later today. Thanks!

waclawjacek commented on plugins/woocommerce/includes/shipping/flat-rate/class-wc-shipping-flat-rate.php:
```diff
@@ -282,6 +282,12 @@ public function sanitize_cost( $value ) {
 		$value = wp_kses_post( trim( wp_unslash( $value ) ) );
 		$value = str_replace( array( get_woocommerce_currency_symbol(), html_entity_decode( get_woocommerce_currency_symbol() ) ), '', $value );
+
+		$contains_shortcodes = false !== strpos( $value, '[' ) || false !== strpos( $value, ']' );
+		if ( ! $contains_shortcodes ) {
+			$value = \Automattic\WooCommerce\Utilities\NumberUtil::sanitize_cost_in_current_locale( $value );
+		}
```
Do y'all think we should follow up with a PR to also handle cost input containing shortcodes?

psealock replied:
>Do y'all think we should follow up with a PR to also handle cost input containing shortcodes?

Sanitization and validations of formulas is an important thing, but I'm not certain we should expand the scope of this effort right now.

waclawjacek replied:
Oh yeah! I mean in another issue.

rjchow replied:
Yea I'll be writing up a separate issue for that - I'm thinking to specify a grammar and then implement a proper parser

waclawjacek commented on plugins/woocommerce/src/Utilities/NumberUtil.php:
```diff
@@ -49,4 +53,66 @@ public static function array_sum( array $arr ): float {
+	 * Sanitize a cost value based on the current locale decimal and thousand separators.
+	 *
+	 * @param string $value               The value to sanitize.
+	 * @return string The sanitized value, or empty string if invalid.
```
If `$value` is invalid, it will throw an exception.

The suggestion below also fixes indentation.

```suggestion
	 * @return string                     The sanitized value.
```

waclawjacek commented on plugins/woocommerce/src/Utilities/NumberUtil.php:
```diff
@@ -49,4 +53,66 @@ public static function array_sum( array $arr ): float {
+			(
+				// Check that decimal separator appears after thousand separator if both exist.
+				str_contains( $value, $thousand_separator ) &&
+				str_contains( $value, $decimal_separator ) &&
```
`str_contains` is only available in PHP 8+.

```suggestion
				false !== strpos( $value, $thousand_separator ) &&
				false !== strpos( $value, $decimal_separator ) &&
```

waclawjacek commented on plugins/woocommerce/src/Utilities/NumberUtil.php:
```diff
@@ -49,4 +53,66 @@ public static function array_sum( array $arr ): float {
+		$value = str_replace( $currency_symbol_variations, '', $value );
+
+		$allowed_characters_regex = sprintf(
+			'/^[0-9\%s\%s]*$/',
```
I am looking at a function called [`wc_format_decimal()`](https://github.com/woocommerce/woocommerce/blob/67407d21886cceca10d0b7cfe429295b461a7d4f/plugins/woocommerce/includes/wc-formatting-functions.php#L288) which recognizes more decimal separators than this method currently does.

It calls `localeconv()` and gets `decimal_point` and `mon_decimal_point` ("mon" stands for "monetary") from it, therefore also allowing the current locale's decimal separators and not only what is set in WC settings.

It also does some other things around sanitizing the value.

Do you think we should be using that function here, or doing some of what it is doing?

I suppose if we have the message explaining what separators are accepted then it should be OK but just wanted to bring this up in case I am missing something.

rjchow replied:
localeconv() would get the server's locale though - and I think in most cases that probably won't be the user's locale

psealock replied:
>It also does some other things around sanitizing the value.

In addition to getting the decimal point from server locale in and the one in Settings this function will convert multiple dots to just one. Otherwise, its the same as we already have, except it does not handle thousands separators.

I think its better to just add in consolidating multiple dots to just one, if applicable, than to use this function in addition to separately handling thousands separators.

psealock committed:
Update plugins/woocommerce/src/Utilities/NumberUtil.php

Co-authored-by: Wacław Jacek <waclaw.jacek@automattic.com>

psealock committed:
Update plugins/woocommerce/src/Utilities/NumberUtil.php

Co-authored-by: Wacław Jacek <waclaw.jacek@automattic.com>

rjchow committed:
update error message

waclawjacek commented on plugins/woocommerce/src/Utilities/NumberUtil.php:
```diff
@@ -76,8 +76,8 @@ public static function sanitize_cost_in_current_locale( $value ): string {
-			/* translators: %s: Invalid value that was input by the user */
-			throw new \InvalidArgumentException( sprintf( esc_html__( '%s is not a valid numeric value. Allowed characters are numbers and the thousand and decimal separators.', 'woocommerce' ), esc_html( $value ) ) );
+			/* translators: %1$s: Invalid value that was input by the user, %2$s: thousand separator, %3$s: decimal separator */
+			throw new \InvalidArgumentException( sprintf( esc_html__( '%1$s is not a valid numeric value. Allowed characters are numbers, and the thousand (%2$s) and decimal (%3$s) separators.', 'woocommerce' ), esc_html( $value ), esc_html( wc_get_price_thousand_separator() ), esc_html( wc_get_price_decimal_separator() ) ) );
```
Not a required change:

How about this instead? Same thing, just less `esc_html`s:

```suggestion
			throw new \InvalidArgumentException(
				esc_html(
					sprintf(
						__( '%1$s is not a valid numeric value. Allowed characters are numbers, and the thousand (%2$s) and decimal (%3$s) separators.', 'woocommerce' ),
						$value,
						wc_get_price_thousand_separator(),
						wc_get_price_decimal_separator()
					)
				)
			);
```

waclawjacek commented:
:eyes:

waclawjacek commented on packages/js/currency/src/utils.tsx:
```diff
@@ -406,3 +406,100 @@ export function getCurrencyData() {
+		throw new Error( 'Invalid decimal separator' );
+	}
+
+	const unformattedValue = inputNumber
+		.replace(
+			new RegExp( escapeRegExp( config.thousandSeparator ), 'g' ),
+			''
+		)
+		.replace( config.decimalSeparator, '.' );
```
This now creates a conflict - the backend (`src/Utilities/NumberUtil.php`) expects a localized number, and the front-end sends an unformatted value.

A way to test this is to set the thousand separator to ` ` and the decimal separator to `,`, then providing this as the price: `1 234,12`

psealock replied:
Yikes, we really can't unlocalize twice: once on client and again on server - Good catch and thanks for bringing it up. 

On the server, what if we check to see if the string is in fact a legit number with `is_numeric` before unlocalizing?

If we don't do that I see a couple options:

1. unlocalize on the client only.
2. unlocalize on the server only.
3. Pass a parameter to say whats being sent (localized/unlocalized) and proceed accordingly.
4. Go back to `<input type="number" ... />` before all this fanciness started.

Logically 2. makes sense, but with the system designed for numbers we'd need to check to make sure the API isn't used elsewhere in the app or by extensions. I think the `is_numeric` check is the way to go.
```

rjchow replied:
_grumble grumble_ why is this so hard 🥲

How about - on the backend:

1. if it has anything other than a single decimal point, attempt to delocalise according to localisation rules
2. if it has a single decimal point && is_numeric, accept it as a number (because is_numeric is slightly looser, [1](https://www.php.net/manual/en/function.is-numeric.php), [2](https://www.php.net/manual/en/language.types.numeric-strings.php))
3. else: error

psealock replied:
Yippie, sounds good to me

rjchow replied:
@waclawjacek does the [commit](https://github.com/woocommerce/woocommerce/pull/54181/commits/5eb5e149fdd1dace6658749dcfb3d4646d917e2c) I pushed look ok to you? I'd like to get this into a 9.7.0 CFE soon

psealock committed:
Update plugins/woocommerce/src/Utilities/NumberUtil.php

Co-authored-by: Wacław Jacek <waclaw.jacek@automattic.com>

rjchow committed:
fixed case where standard decimals were being rejected as it was expecting a localised number

ismaeldcom commented:
If I'm not wrong this PR should also fix #54768 and #50717

waclawjacek commented:
:eyes:

kallehauge commented on plugins/woocommerce/src/Utilities/NumberUtil.php:
```diff
@@ -49,4 +53,85 @@ public static function array_sum( array $arr ): float {
+				esc_html(
+					sprintf(
+						/* translators: %1$s: Invalid value that was input by the user, %2$s: thousand separator, %3$s: decimal separator */
+						__( '%1$s is not a valid numeric value. Allowed characters are numbers, and the thousand (%2$s) and decimal (%3$s) separators.', 'woocommerce' ),
```
I really like the new inclusion of expected separators!

![Screenshot 2025-02-13 at 23 34 23](https://github.com/user-attachments/assets/bfceaccd-555c-43d3-8e5f-3004214d6024)

**[Non-blocking, opinionated] UX suggestion**

I'm 99% sure this is outside scope, but might it be a better experience to show the error inside the edit modal? The current experience closes the modal on save to then surface the error message. 

It's quite niche, but if I had loads of shipping methods, then it might be annoying trying to find the one that failed because I was in a groove setting up or something (again, very niche!)

Alternatively, then we could also include the name of the shipping method in the error message? It's part of the `woocommerce_shipping_zone_methods_save_settings` response, so it should be doable?

_I want to recognise that the display of the error happens elsewhere which is why I consider it outside scope, but sharing it just for the sake of it anyway_

```json
{
    "success": true,
    "data": {
        "zone_id": 1,
        "zone_name": "US",
        "methods": {
            "1": {
                "min_amount": "5",
                "requires": "min_amount",
                "id": "free_shipping",
                "method_title": "Free shipping",
                "method_description": "<p>Free shipping is a special method which can be triggered with coupons and minimum spends.<\/p>\n",
                "title": "Free shipping",
                ...
            }
        },
        "errors": [
            "5_000 is not a valid numeric value. Allowed characters are numbers, and the thousand (,) and decimal (.) separators."
        ]
    }
}
```

rjchow replied:
I did consider it out of concern for the UX, but yea decided it was out of scope especially because it's written in very old school JS that I'm not confident with making big changes in, and we'll probably be reworking this with the settings project

kallehauge commented on plugins/woocommerce/src/Utilities/NumberUtil.php:
```diff
@@ -49,4 +53,85 @@ public static function array_sum( array $arr ): float {
+
+		/**
+		 * For context, as of 2025.
+		 * The full set of thousands separators is PERIOD, COMMA, SPACE, APOSTROPHE.
```
ℹ️  I tested digit groupings just for the sake of it since it's technically supported by [PHP (7.4+)](https://www.php.net/manual/en/language.types.integer.php) which is responsible for validating our numeric values and it's being rejected as expected (I say "as expected" since it might be a common programming pattern but has nothing to do with $$$).

_Context: PHP, JS, and some other programming languages allow underscores (`_`) to do digit groupings. Example: `1_000_000` means `1000000` and the underscores are purely for presentation as a way to more easily read numbers._
_You can open the JS dev console and enter `1_000*10` which will return `10000`_.

rjchow replied:
I think I'd rather not support this 😅 It's complex enough handling just the locale-specifics

kallehauge replied:
Oh yeah, totally agree! My intention was to share that we're not, unintentionally, supporting underscores just because the programming language supports it - sorry about any confusion! 😄

kallehauge commented on plugins/woocommerce/src/Utilities/NumberUtil.php:
```diff
@@ -49,4 +53,85 @@ public static function array_sum( array $arr ): float {
+				esc_html(
+					sprintf(
+						/* translators: %1$s: Invalid value that was input by the user, %2$s: thousand separator, %3$s: decimal separator */
+						__( '%1$s is not a valid numeric value. Allowed characters are numbers, and the thousand (%2$s) and decimal (%3$s) separators.', 'woocommerce' ),
```
I believe Matt generally pushes for Oxford commas which (AFAIK) just means we should add a `,` before `and` in this case because it's "a list" 😄 

```suggestion
						__( '%1$s is not a valid numeric value. Allowed characters are numbers, the thousand (%2$s), and decimal (%3$s) separators.', 'woocommerce' ),
```

kallehauge commented on plugins/woocommerce/src/Utilities/NumberUtil.php:
```diff
@@ -49,4 +53,85 @@ public static function array_sum( array $arr ): float {
+				strpos( $value, $decimal_separator ) <= strpos( $value, $thousand_separator )
+			)
+		) {
+			/* translators: %s: Invalid value that was input by the user */
+			throw new \InvalidArgumentException( sprintf( esc_html__( '%s is not a valid numeric value: there should be one decimal separator and it has to be after the thousands separator.', 'woocommerce' ), esc_html( $value ) ) );
```
[Non-blocking, opinionated] Suggestion to keep our formatting the same and we've approved this once already with the above `InvalidArgumentException` (I personally prefer this as well since I read it better - just to share my two cents between the two):

```suggestion
			throw new \InvalidArgumentException(
				esc_html(
					sprintf(
						/* translators: %s: Invalid value that was input by the user */
						__( '%s is not a valid numeric value: there should be one decimal separator and it has to be after the thousands separator.', 'woocommerce' ),
						$value
					)
				)
			);
```

kallehauge commented on plugins/woocommerce/src/Utilities/NumberUtil.php:
```diff
@@ -49,4 +53,85 @@ public static function array_sum( array $arr ): float {
+
+		if ( $value && ! is_numeric( $value ) ) {
+			/* translators: %s: Invalid value that was input by the user */
+			throw new \InvalidArgumentException( sprintf( esc_html__( '%s is not a valid numeric value', 'woocommerce' ), esc_html( $value ) ) );
```
[Non-blocking, opinionated] Suggestion to keep our formatting the same and we've approved this once already with the above `InvalidArgumentException` (I personally prefer this as well since I read it better - just to share my two cents between the two):

```suggestion
			throw new \InvalidArgumentException(
				esc_html(
					sprintf(
						/* translators: %s: Invalid value that was input by the user */
						__( '%s is not a valid numeric value.', 'woocommerce' ),
						$value
					)
				)
			);
```

kallehauge commented on packages/js/currency/src/test/utils.ts:
```diff
@@ -0,0 +1,285 @@
+		);
+	} );
+
+	it( 'should not format incorrectly formatted numbers according to current config', () => {
```
[Non-blocking; opinionated]: Since we've already been concerned about failing silently, should we throw an error for incorrect thousand separator placements? 

I'm working of a personally biased understanding of how I would work with WC, but if I add a thousand separator, then I'd expect it to be in the correct location, so if I wrote `7,50` then I'd prefer to get an error saying it doesn't look like an expect format (which should probably have been `7,500`).
That being said then I could also have mistyped `75,00` which should be `7500` so ignoring it would be correct in this case; but is that enough to counterweight allowing the other case?

rjchow replied:
This function actually formats the value retrieved from the backend, for display in the frontend - I think it's correct to ignore incorrectly formatted numbers so that it throws an error when the user saves it. We don't currently have a routine that displays an error thrown where this function is called 🥲

This is out of consideration that we might have mal-formatted numbers saved prior to this PR, which should not occur going forward.

kallehauge commented on plugins/woocommerce/src/Utilities/NumberUtil.php:
```diff
@@ -49,4 +53,85 @@ public static function array_sum( array $arr ): float {
+		$value = str_replace( $currency_symbol_variations, '', $value );
+
+		// Count the number of decimal points.
+		$decimal_point_count = substr_count( $value, '.' );
```
> @waclawjacek does the [commit](https://github.com/woocommerce/woocommerce/pull/54181/commits/5eb5e149fdd1dace6658749dcfb3d4646d917e2c) I pushed look ok to you? I'd like to get this into a 9.7.0 CFE soon

@rjchow I'm not entirely sure here. If `.` has been defined as the thousand separator, wouldn't we be making an incorrect assumption by hardcoding the decimal pointer?
```suggestion
		$decimal_point_count = substr_count( $value, wc_get_price_decimal_separator() );
```

rjchow replied:
This is hardcoded because we expect the frontend to produce standard PHP numbers with only `.` for decimal separators, after this PR is merged. This is a way of cleaning up the database going forward so that we only have PHP numbers in the database instead of locale-specific numbers

kallehauge commented on plugins/woocommerce/src/Utilities/NumberUtil.php:
```diff
@@ -49,4 +53,85 @@ public static function array_sum( array $arr ): float {
+		$decimal_point_count = substr_count( $value, '.' );
+
+		// If it's a standard decimal number (single decimal point and is_numeric), accept it directly. This could be in the case where the frontend has de-localised the value.
+		// We check for the decimal point count because is_numeric is looser than that.
```
Should we make the comment more specific and explain how it's looser and the consequences that makes us do custom checks?

rjchow replied:
Good callout because I forgot how it's looser by now 🤣 I'll update the comment

rjchow replied:
```suggestion
		// We check for the decimal point count in addition to using is_numeric.
		// This is because is_numeric is much looser and accepts non-base10 numbers as well as 'e' to demarcate exponents.
```

kallehauge commented:
**Existing values**

Just making sure that it's expected behaviour 😄 

The test instructions doesn't say anything about values that has already been stored to the database using `trunk`. Example: minimum amount for free shipping being `(string) 100,00` and the store uses `,` as decimal separators:

```php
a:4:{s:5:"title";s:13:"Free shipping";s:8:"requires";s:10:"min_amount";s:10:"min_amount";s:6:"100,00";s:16:"ignore_discounts";s:2:"no";}
``` 

But if I inject this into my database and use this branch, then checkout behaves as expected (read: Free Shipping is only available when the cart is >100) 😄

rjchow committed:
Apply suggestions from code review

Co-authored-by: André Kallehauge <3846700+kallehauge@users.noreply.github.com>

rjchow committed:
fixed test error expectation

rjchow committed:
fix test error expectation take 2

rjchow committed:
Merge branch 'trunk' into fix/shipping-method-values-unformatted

psealock commented:
Thanks for the thoughtful review @kallehauge !

>Just making sure that it's expected behaviour 😄

Yes I think so. Although were not explicitly handling these cases, nor do I think we should because we'd be making stabs in the dark to correct things. Targeted outreach will be better. The error generally affected stores who operate currencies that require large denominations, such as shipping an item for 20,000 pesos. Of course thats a generalization, but a blog post would be useful to alert merchants the potentially serious error has been fixed.

Otherwise, was your testing sufficient to offer enough certainty to deem this fix robust?

kallehauge reviewed (approved):
>> Just making sure that it's expected behaviour 😄
>
>Yes I think so. Although were not explicitly handling these cases, nor do I think we should because we'd be making stabs in the dark to correct things. Targeted outreach will be better. The error generally affected stores who operate currencies that require large denominations, such as shipping an item for 20,000 pesos. Of course thats a generalization, but a blog post would be useful to alert merchants the potentially serious error has been fixed.

Makes sense! 👍 

> Otherwise, was your testing sufficient to offer enough certainty to deem this fix robust?

@psealock, I'd say so! It certainly held up to the different tests I threw at it 😄

rjchow commented:
Thanks both @waclawjacek and @kallehauge for providing infinitely valuable feedback on this PR! It was a team effort!
